File: RecyclerViewPagination/app/src/main/java/com/kolon/recyclerviewpagination/MainActivity.java
package com.kolon.recyclerviewpagination;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.View;
import android.widget.RelativeLayout;
import com.kolon.recyclerviewpagination.log.TimberLog;
import java.util.ArrayList;
import timber.log.Timber;

public class MainActivity extends AppCompatActivity {

    private RecyclerView recyclerView;
    private RecyclerViewAdapter adapter;
    private RelativeLayout progressBar;

    private ArrayList<String> pagingDataList = new ArrayList<>();
    private ArrayList<String> itemDataList = new ArrayList<>();

    private boolean isLoading = false;

    private final int ORIGIN_DATA_SIZE = 100;
    private int initialSize = 10;
    private int pageSize = 20;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        this.recyclerView = findViewById(R.id.recyclerView1);
        this.progressBar = findViewById(R.id.main_progressbar);

        // TimberLog 적용
        Timber.plant(new TimberLog());
        Timber.i("START TIMBER JAVA!!!");

        initData();
        populateData();
        initAdapter();
        initScrollListener();
    }

    // paging 처리할 임시 datat생
    public void initData() {
        for (int i = 0; i < ORIGIN_DATA_SIZE; i++) {
            itemDataList.add("ITEM :: " + String.valueOf(i + 1));
        }
    }

    // 초기 PaingData setting생성, 처음 10개만 보여준다.
    private void populateData() {
        for (int i = 0; i < this.initialSize; i++) {
            pagingDataList.add(itemDataList.get(i));
        }
    }

    private void initAdapter() {
        adapter = new RecyclerViewAdapter(this, pagingDataList);
        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(layoutManager);
    }


    // 리사이클러 뷰 리스
    private void initScrollListener() {
        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
            }

            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);

                LinearLayoutManager layoutManager = (LinearLayoutManager) recyclerView.getLayoutManager();

                int currentLastItem = layoutManager.findLastCompletelyVisibleItemPosition() + 1;
                int currentDataSize = pagingDataList.size();

                if (!isLoading) {
                    if (layoutManager != null && currentLastItem == currentDataSize) {
                        //리스트 마지막에 도달하면, Progressbar보여준다.
                        progressBar.setVisibility(View.VISIBLE);
                        loadMore();
                        isLoading = true;
                    }
                }
            }
        });
    }

    private void loadMore() {
        new Handler(Looper.getMainLooper()).postDelayed(new Runnable() {
            @Override
            public void run() {

                pagingDataList.add(null);
                adapter.notifyItemInserted(pagingDataList.size() - 1);

                pagingDataList.remove(pagingDataList.size() - 1);
                int scrollPosition = pagingDataList.size();

                adapter.notifyItemRemoved(scrollPosition);

                int currentSize = scrollPosition;
                int nextLimit = currentSize + pageSize;  // 20개씩 늘어남.

                if (nextLimit > itemDataList.size()) {
                    nextLimit = itemDataList.size();
                }

                while (currentSize - 1 < nextLimit) {
                    pagingDataList.add("Item " + currentSize);
                    currentSize++;
                }

                adapter.notifyDataSetChanged();
                isLoading = false;

                progressBar.setVisibility(View.GONE);
            }
        }, 2000);
    }
}

File: RecyclerView_Kotiln/app/src/main/java/com/kolon/recyclerview_kotiln/MainActivity.kt
package com.kolon.recyclerview_kotiln

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.recyclerview.widget.LinearLayoutManager
import com.kolon.recyclerview_kotiln.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {

    // 전역 변수로 바인딩 객체 선언
    private var mBinding: ActivityMainBinding? = null

    // 매번 null check를 하지 않도록 바인딩 변수 선언
    private val binding get() = mBinding!!


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //setContentView(R.layout.activity_main)

        mBinding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)


        val profileList = arrayListOf<Profiles>(
            Profiles(R.drawable.man, "고진권", 29, "코오롱베니트 안드로이드 앱 개발자"),
            Profiles(R.drawable.woman, "박서영", 25, "코오롱베니트 서버 개발자"),
            Profiles(R.drawable.man, "박성준", 29, "농협 은행 개발자"),
            Profiles(R.drawable.man, "박지상", 28, "스타트업 안드로이드 앱 개발자"),
            Profiles(R.drawable.man, "정구일", 29, "수협은행 안드로이드 앱 개발자"),
            Profiles(R.drawable.man, "김형택", 28, "넷바블 스프링 개발자"),
            Profiles(R.drawable.man, "고진권", 30, "안드로이드 앱 개발자"),
            Profiles(R.drawable.man, "고진권", 31, "안드로이드 앱 개발자"),
            Profiles(R.drawable.man, "고진권", 32, "안드로이드 앱 개발자"),
            Profiles(R.drawable.woman, "고진권", 33, "안드로이드 앱 개발자"),
            Profiles(R.drawable.man, "고진권", 34, "안드로이드 앱 개발자"),
            Profiles(R.drawable.woman, "고진권", 35, "안드로이드 앱 개발자"),
            Profiles(R.drawable.man, "고진권", 36, "안드로이드 앱 개발자"),
            Profiles(R.drawable.man, "고진권", 37, "안드로이드 앱 개발자")
        )

        binding.rvProfile.layoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
        binding.rvProfile.setHasFixedSize(true) // 리사이클러 뷰에 대한 성능 개선 방안

        // 이제 본격적으로 우리가 만든 adapter를 최초 생성해서 연결
        binding.rvProfile.adapter = ProfileAdapter(profileList)
    }
}